### ISSUE TITLE
Add a daemon mode to prevent duplicate processes

### ISSUE THREAD

**Issue by mrjpaxton** [foresto/joystickwake#13]
Hey, I know this project is pretty small, but I just want to say that I really appreciate this, because nobody else seems to be doing anything else like this right now.

This is a suggestion rather than a bugfix, but I was wondering if we can figure out how to add a daemon mode (`-d` flag) to this program, because when Linux desktops or WMs run it, or when switching between Linux desktops, it is possible to fork more than one process.

It will probably require figuring out how to enable and run a process PID runtime in either `/run` or `/run/user/...` or as otherwise specified with `$XDG_RUNTIME_DIR` with Python 3 and make sure no duplicates are run as long as the `-d` flag is specified. I'm not sure how the "proper way" of doing this stuff is implemented.

I have limited experience with Python 3, but I know how to read some of the syntax, I don't have much experience with Git however, just only some of the bare bones knowledge, so maybe I can only help a little bit.

It may also be possible to encapsulate the process in a Systemd unit file. However, this would only be compatible with systems running Systemd, so it would be limited to that. And besides, Systemd might be able to take advantage of a `-d` daemon mode anyway, so this is probably a higher priority.

Let me know the best way I can help contribute to this goal if you agree that this is good idea, or if any ideas need refining.

**Comment by foresto** [comment]
I know how to make a daemon, but I've never seen a situation that calls for it with joystickwake.

If you're getting duplicate instances of it, how did you install it and how are you launching it?

The best way is to have it launched by a session startup .desktop file (in $HOME/.config/autostart) or script (like $HOME/.xsession). Either of those will start it only once, when you log in to the desktop session, and Joystickwake will quit on its own when you log out.

In case you hadn't noticed, this repo contains a .desktop file ready for use, assuming joystickwake is installed in a directory that's in your $PATH.

**Comment by mrjpaxton** [comment]
I guess I have to admit I prefer keeping things extremely lightweight, but also functional. I don't use  display manager, I login to a tty to run `startx` or a Wayland desktop.

I am using `exec` commands in both i3 and Sway WM. See this for details - https://i3wm.org/docs/userguide.html#exec (This largely applies to Sway as well)

It is designed to autoexec a process when it starts up. In i3's case, I use `picom -b`, since it already has a daemon mode. I could also use `~/.xinitrc`, but there are some cases where I don't need `picom` for certain other desktops.

In the case of Sway, well... There is no `~/.xinitrc`. I think `joystickwake` is forked and when Sway is closed, the process still remains, same for i3.

I guess a workaround would be to make sure `pkill joystickwake` runs when these WM desktops close. I wonder how I could do that...

**Comment by foresto** [comment]
> It is designed to autoexec a process when it starts up. In i3's case, I use `picom -b`, since it already has a daemon mode. I could also use `~/.xinitrc`, but there are some cases where I don't need `picom` for certain other desktops.

Can't you use ~/.xsession or [whatever equivalent](https://manpages.org/xsession/5) is supported by your distro's startx command?

If not, can't that autoexec process be a simple script that launches joystickwake and (conditionally) picom?

> In the case of Sway, well... There is no `~/.xinitrc`. 

I don't use Sway, but I would expect whatever method you use to launch it to have a facility for running programs at session start, just as startx has ~/.xinitrc and ~/.xsession. It's probably worth finding out what that facility is.

> I think `joystickwake` is forked and when Sway is closed, the process still remains, same for i3.

Forked by what? If Joystickwake can't find an X display, it exits when its parent process (which forked it) exits.

If you really must have a daemon, you could always use [daemonize](https://software.clapper.org/daemonize/), which I think is available in most distros, or write a unit file for systemd if you're using that.

**Comment by foresto** [comment]
Closing because the reporter hasn't shown a convincing need for built-in daemon support, nor identified a common use case that actually yields duplicate processes.

For what it's worth, I get the appeal of traditional startx and lightweight GUI sessions. I just don't see how adding daemon code to joystickwake would accomplish anything that isn't done better externally.

**Comment by nonetrix** [comment]
Made a SystemD service with NixOS asking Claude. Seems to work so far besides what I think is mostly a Hyprland bug, hopefully helps others :-)
```
{
  inputs,
  pkgs,
  ...
}: let
  global = import ../../../global-var.nix;
in {
  home-manager.users.${global.username} = { pkgs, ... }: {
    systemd.user.services.joystickwake = {
      Unit = {
        Description = "Joystick-aware screen waker for games";
        After = [ "graphical-session.target" ];
        PartOf = [ "graphical-session.target" ];
      };

      Install = {
        WantedBy = [ "graphical-session.target" ];
      };

      Service = {
        Environment = "DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/%U/bus";
        ExecStart = "${pkgs.writeShellScript "start-joystickwake" ''
          #!/run/current-system/sw/bin/bash
          
          # Use busctl to inhibit screen saver
          exec ${pkgs.joystickwake}/bin/joystickwake \
            --cooldown 15 \
            --command "busctl --user call org.freedesktop.ScreenSaver /org/freedesktop/ScreenSaver org.freedesktop.ScreenSaver Inhibit ss 'test' 'foobar'"
        ''}"
        ;
        
        # Restart on failure with a delay
        Restart = "on-failure";
        RestartSec = 30;
        
        # Nice-to-have settings
        Type = "simple";
        PrivateTmp = true;
        NoNewPrivileges = true;
      };
    };
  };
}
```
